charon/pkgs/maven.py

```diff
@@ -497,6 +613,199 @@ def _scan_paths(files_root: str, ignore_patterns: List[str],
+                local_archetypes = _parse_archetypes(f.read())
+
+            if len(local_archetypes) < 1:
+                # If there are no local archetypes in the catalog,
+                # there's nothing to do.
+                logger.warning(
+                    "No archetypes found in local archetype-catalog.xml, "
+                    "even though the file exists! Skipping."
+                )
+                return 0
+
+            else:
+                # Read the archetypes from the bucket so we can do a merge / un-merge
+                remote_xml = s3.read_file_content(bucket, remote)
+                remote_archetypes = _parse_archetypes(remote_xml)
+
+                if len(remote_archetypes) < 1:
+                    # Nothing in the bucket. Clear out this empty file.
+                    __gen_all_digest_files(local)
+                    return -1
+
+                else:
+                    # If we're deleting, un-merge the local archetypes from
+                    # the remote ones.
+                    #
+                    # NOTE: The ONLY reason we can get away with this kind of
+                    # naive un-merge is that products only bother to publish
+                    # archetypes for their own direct users. If they publish
+                    # an archetype, it's only for use with their product.
+                    # Therefore, if we rollback that product, the archetypes
+                    # they reference shouldn't be useful anymore.
+                    for la in local_archetypes:
+                        if la in remote_archetypes:
+                            remote_archetypes.remove(la)
+
+                    if len(remote_archetypes) < 1:
+                        # If there are no remote archetypes left after removing
+                        # ours DELETE the bucket catalog.
+                        __gen_all_digest_files(local)
+                        return -1
+                    else:
+                        # Re-render the result of our archetype un-merge to the
+                        # local file, in preparation for upload.
+                        with open(local, 'wb') as f:
+                            content = MavenArchetypeCatalog(remote_archetypes)\
+                                .generate_meta_file_content()
+                            try:
+                                overwrite_file(local, content)
+                            except FileNotFoundError as e:
+                                logger.error(
+                                    "Error: Can not create file %s because of some missing folders",
+                                    local,
+                                )
+                                raise e
+                        __gen_all_digest_files(local)
+                        return 1
+
+    return 0
+
+
+def _generate_upload_archetype_catalog(
+        s3: S3Client, bucket: str,
+        root: str, prefix: str = None
+) -> bool:
```

Comment:
I think we need to change this to `-> Tuple[bool, List[str]]`, so we can account for the digest files that might need to be uploaded alongside the archetype-catalog.xml file itself.